### PR TITLE
Add JUnitXML support

### PR DESCRIPTION
**Pull request by scottclowe**
Add support for generating an XML output file in the format used by JUnit.

There are several CI servers (Jenkins, Shippable, Bamboo) which read in JUnit style XML to display the test results.

I also added a tic/toc to time the tests and include the time in the XML output.
And I changed `moxunit_utility_elem2str` to give more informative outputs and not use `<` or `>`, since these were breaking the XML format.

A usage example is [available here](https://app.shippable.com/builds/566339ff358b060c00f68c5f). Just click on the Tests tab to see the test details.

Note: This pull request includes the change from my pull request in #3, because otherwise the build would fail.

### CODE REVIEW

**Comment by nno** [comment]
Thanks a lot for the PR - it would be nice to add XML support.

It is clear that you spent quite some time on this, and the code is generally well readable - your work is really appreciated. However, I wonder if XML support can be added using a more simple approach: 

All reporting is done through the `@MoxUnitTestResult` class, and it would seem sufficient to output either in (human-readable) text format or in XML. What would you think about adding an extra argument to `@MOxUnitTestResult/MOxUnitTestResult`, for example "output_format", which can be either "text" or "xml"? The format can then be stored inside the `@MOxUnitTestResult` object, and then only `report`, `disp`, and `show_non_successes` need updating. Actually if XML only needs reports for each test but no summaries, then the `disp` and `show_non_successes` can immediately return (without providing output) when using XML output. The `report` function could have two subfunctions, `report_text` and `report_xml`, where the latter could use code you already wrote. In that way there is no need to store additional XML output in the test result class.
`moxunit_run_tests` can be updated, for example by adding such a 'output_format' option, which would be passed to the `@MOxUnitTestResult` constructor.

What do you think? Do you think this would make things simpler?

Also, I like the idea of measuring the amount of time each test took - but I would prefer to do this in a separate PR.

File: MOxUnit/@MOxUnitFunctionHandleTestCase/run.m [hunk 1]
```diff
@@ -16,6 +16,7 @@
 %
 % NNO 2015
 
+    start_tic = tic;
```
Comment (nno) [inline MOxUnit/@MOxUnitFunctionHandleTestCase/run.m]: I like the idea of measuring time - but it may be better to do this in a separate PR?
Also can the use of tic/toc be avoided? It could override any tic/toc use by the user. How about using clock() instead?

File: MOxUnit/@MOxUnitTestResult/MOxUnitTestResult.m [hunk 2]
```diff
@@ -34,5 +34,7 @@
     s.skips=cell(0);
     s.successes=cell(0);
     s.testsRun=0;
+    s.duration=0;
+    s.xml_reports=cell(0);
```
Comment (nno) [inline MOxUnit/@MOxUnitTestResult/MOxUnitTestResult.m]: - this would add some duplication in the information represented.
- using this approach, XML results are not updated after each test (as is currently the case) but only after all tests have been run.

Comment (nno) [inline MOxUnit/@MOxUnitTestResult/MOxUnitTestResult.m]: If XML needs a header to be printed before any test result, then the report function should, when output is set to XML, check to see if any tests has already been run - and if not, print the XML header.

File: MOxUnit/@MOxUnitTestResult/report_xml.m [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    % Set up the XML node for the testcase.
+    % They all start the same way.
+    x = '<testcase classname="%s" file="%s" name="%s" time="%.9f"';
+
+    % We need to include the classname although we don't rightly have
+    % a classname to give. Instead, we use the path to the file.
+    classpath = get(t,'location');
+    % Remove out the path to the testing directory
+    classpath = strrep(classpath, pwd, '');
```
Comment (nno) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: - Why remove the path to the testing directory?
- What if pwd has multiple occurrences in classpath (if that's possible)?

File: MOxUnit/@MOxUnitFunctionHandleTestCase/run.m (hunk 1 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitFunctionHandleTestCase/run.m]: I was going to split the PR, but then the XML code became contingent on the timing being available. I can still split them if you prefer, but the XML code will then need another PR (or there will be XML code in the timing PR).

Using `tic` is fine because when you set a variable equal to a tic it goes into a different timer than the global last tic.

Demonstration:
```matlab
disp('start tic test');
tic;
pause(1);
toc;
inner_tic = tic;
pause(1.5);
toc(inner_tic);
toc;
disp('end tic test');
```
Output:
```
start tic test
Elapsed time is 1.004555 seconds.
Elapsed time is 1.507640 seconds.
Elapsed time is 2.522680 seconds.
end tic test
```

I would be happy with whatever gives the best result. I don't know if tic is more accurate than clock, but tic is certainly easier to use.

File: MOxUnit/@MOxUnitTestResult/report_xml.m [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+%                   escaped.
+%
+% SCL 2015
+
+    % We check to make sure the character before the '<' isn't a '\' or
+    % the start of the string. If it isn't, we capture until the next
+    % non-escaped '>' character and remove the whole chunk. However,
+    % we need to include the leading non-'\' character (the first group)
+    % in the replacement output.
+    string = regexprep(string , '(^|[^\\])(<.*?[^\\]>)', '$1');
```
Comment (nno) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: In the second `[^\\]` occurrence, how could it match the start of the string? Can `^` be omitted there?

File: MOxUnit/@MOxUnitTestResult/report_xml.m [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+%   string          The string with any XML tags removed. Any escaped
+%                   characters ('\<', '\>') remain as they are - still
+%                   escaped.
+%
+% SCL 2015
+
+    % Trim off any newlines from the start of the string
+    % NB: sprintf('\n')==10
+    string = strtrim(string);
+    start_index = find(string(2:end)~=10, 1);
```
Comment (nno) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: This seems platform dependent... How about using sprintf('\n') instead?

File: MOxUnit/assertEqual.m [hunk 6]
```diff
@@ -26,14 +26,14 @@ function assertEqual(a, b, message)
...
-    if ~isequal(size(a), size(b))
-        whatswrong='inputs are not of the same size';
-        error_id='assertEqual:nonEqual';
-
-    elseif ~isequal(class(a), class(b))
+    if ~isequal(class(a), class(b))
         whatswrong='inputs are not of the same class';
         error_id='assertEqual:classNotEqual';
 
+    elseif ~isequal(size(a), size(b))
```
Comment (nno) [inline MOxUnit/assertEqual.m]: Why the change in order?

File: MOxUnit/moxunit_runtests.m [hunk 7]
```diff
@@ -43,7 +43,7 @@
 % NNO Jan 2014
 
 
-    [verbosity, filenames, fid]=get_params(varargin{:});
+    [verbosity, filenames, fid, junitxml]=get_params(varargin{:});
```
Comment (nno) [inline MOxUnit/moxunit_runtests.m]: As mentioned earlier: how about adding a 'output_format' argument which can be set to text or xml?

File: MOxUnit/moxunit_runtests.m [hunk 8]
```diff
@@ -142,3 +156,28 @@
         filenames={pwd()};
     end
 
+
+function write_xml(fname, test_result)
+
+    if isempty(fname);
+        % No file to write to, so stop now.
```
Comment (nno) [inline MOxUnit/moxunit_runtests.m]: Is this allowed? Why not throw an error?

File: MOxUnit/moxunit_runtests.m [hunk 9]
```diff
@@ -142,3 +156,28 @@
...
+        return;
+    end
+
+    % Open file to start writing XML report
+    fid=fopen(fname,'w');
+    if fid==-1
```
Comment (nno) [inline MOxUnit/moxunit_runtests.m]: Please use 

file_closer=onCleanup(@()fclose(fid))

File: MOxUnit/moxunit_runtests.m [hunk 10]
```diff
@@ -142,3 +156,28 @@
...
+        error('Could not open file %s for writing', fn);
+    end
+
+    % Start with the XML header
+    fprintf(fid, '<?xml version="1.0" encoding="utf-8"?>');
```
Comment (nno) [inline MOxUnit/moxunit_runtests.m]: When using the TestResult class, the .stream property could be used directly in fprintf there - and output can be written to standard output (fid=1) or to a file.

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 3 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: To the first, because it gives a better user experience and is more compatible with the way classname works for other JUnit XML outputs.

To the second, yes I should change it to only remove from the start. I will do that.

File: MOxUnit/moxunit_util_elem2str.m [hunk 11]
```diff
@@ -18,20 +18,54 @@
...
-    else
-        % Octave does not support evalc
+    % Strings are trivally supported by leaving them as-is
+    if ischar(elem)
+        % Encapsulate in quotation marks
+        elem_str=['''' elem ''''];
+
+    elseif isnumeric(elem)
+        % If the element is numeric, we can use mat2str
+        elem_str=mat2str(elem);
```
Comment (nno) [inline MOxUnit/moxunit_util_elem2str.m]: I would be worried that this creates huge outputs when the input is huge. Already mat2str(eye(1e3)) takes a while to compute and creates a *long* output

File: MOxUnit/moxunit_util_elem2str.m [hunk 12]
```diff
@@ -18,20 +18,54 @@
...
+
+    elseif exist('evalc', 'builtin')
+        % If we have `evalc`, we can just trust the `display` function
```
Comment (nno) [inline MOxUnit/moxunit_util_elem2str.m]: - this could also lead to huge outputs
- it could also raise an error, because we don't have control over how disp is implemented for each class

File: MOxUnit/moxunit_util_elem2str.m [hunk 13]
```diff
@@ -18,20 +18,54 @@
...
+            elem_str = ['{ ' strtrim(elem_str) ' }'];
+        end
+
+    elseif iscell(elem)
+        % Handle cells, without using `evalc`
+        % Just show the size of the cell and indicate the type with
+        % curly braces. This is how MATLAB displays cells in cells.
+        kk = size(elem);
+        xx = repmat('x', 1, 2*length(kk)-1);
+        xx(1:2:end) = arrayfun(@num2str, kk);
```
Comment (nno) [inline MOxUnit/moxunit_util_elem2str.m]: What is any dimension in elem is greater than 9? It would raise an error, no?

File: MOxUnit/moxunit_util_elem2str.m [hunk 14]
```diff
@@ -18,20 +18,54 @@
...
+        elem_str = ['{' xx ' cell}'];
+
+    else
+        % Octave does not support evalc, so we need a work around.
+        % As a fall back, we just show the class of the element
+        elem_str=sprintf('*%s instance*\n',class(elem));
```
Comment (nno) [inline MOxUnit/moxunit_util_elem2str.m]: I would prefer to keep this function simple and stupid, and just use class in all cases. It would be fine with me to remove the '<' and '>' characters. Maybe even try to determine the size (within a try-catch statement)

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 4 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: No. `^` is both start of string and not when used as the first character in a `[]` group.
`[^\\]` == "anything except \"

Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: Just realised this won't capture `<>`... Huumm don't know an easy fix for that.

File: MOxUnit/@MOxUnitFunctionHandleTestCase/run.m (hunk 1 above)
Comment (nno) [inline MOxUnit/@MOxUnitFunctionHandleTestCase/run.m]: Thanks for explaining the var = tic(); syntax, I was not aware of that.
Still, when using tic: does that mean that this function has to store all the start times for every single test? 
Instead, when using `clock`, the `etime` function also returns the time in seconds - so that is about as easy to use a `tic`, isn't it?

File: MOxUnit/moxunit_util_elem2str.m (hunk 13 above)
Comment (scottclowe) [inline MOxUnit/moxunit_util_elem2str.m]: Yes, well spotted.
Replace elseif block with this:
```
xx = strjoin(arrayfun(@num2str, size(elem), 'UniformOutput', false), 'x');
```

File: MOxUnit/moxunit_util_elem2str.m (hunk 11 above)
Comment (scottclowe) [inline MOxUnit/moxunit_util_elem2str.m]: Propose replacing with:
```
elseif isnumeric(elem) && numel(elem) < max_chars
```
And having a different handling for larger numerics, similar to the cell handler?

File: MOxUnit/moxunit_runtests.m (hunk 8 above)
Comment (scottclowe) [inline MOxUnit/moxunit_runtests.m]: It is just because I had the default junitxml input as `''`, so empty is the flag for no output.

However the if catch is also handled when calling `write_xml`, which is unnecessary duplication. So this catch can indeed be removed and potentially be replaced with an error.

File: MOxUnit/assertEqual.m (hunk 6 above)
Comment (scottclowe) [inline MOxUnit/assertEqual.m]: See commit message.
>>>
If the classes are different, this is a more fundamental problem
than the sizes being different and so it is more useful to know
this instead.

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 4 above)
Comment (nno) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: You're right about the '^' as the first character, thanks.
Not sure about an easy fix for `<>` either. It may require some code that just goes over all the characters in a for/while fashion.

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 5 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: [Unicode](https://en.wikipedia.org/wiki/Newline#Unicode) always has LF encoded as 10 in decimal.
So comparing with 10 will be the same as using sprintf('\n') on all systems. I do things this way because `find(_, 1)` is faster than `strfind` --- strfind can't stop after 1.
> LF:    Line Feed, U+000A

File: MOxUnit/assertEqual.m (hunk 6 above)
Comment (nno) [inline MOxUnit/assertEqual.m]: I see your point, but the original Matlab xunit framework checks for equal class first. I prefer to keep it compatible with that when possible. A comment would be useful though to explain the order.

File: MOxUnit/@MOxUnitFunctionHandleTestCase/run.m (hunk 1 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitFunctionHandleTestCase/run.m]: No, it only temporarily stores the start time of the current test in the local variable `start_tic`.
MATLAB just handles `tic` without any output arguments in a special way.

The [help for etime](https://uk.mathworks.com/help/matlab/ref/etime.html) explictly says to use tic and toc to time functions and not etime because the system clock is not reliable.

File: MOxUnit/moxunit_util_elem2str.m (hunk 11 above)
Comment (nno) [inline MOxUnit/moxunit_util_elem2str.m]: I see the point of prettier printing of different datastructures. I really do. Indeed, I've written something like this for another toolbox: https://github.com/CoSMoMVPA/CoSMoMVPA/blob/master/mvpa/cosmo_disp.m

Also I don't have principal objections against some prettier printing of some objects - though the question is where to draw the line, how many special cases on wants to include, etc. 

But can we please discuss this pretty printing functionality separately from this PR (which is about XML output)? Apart from not printing '<' and '>' characters, the two features seem orthogonal.

**Comment by scottclowe** [comment]
I will split the code into 3 PRs
- add xml output support
- time how long functions take
- prettify printing of elements

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 5 above)
Comment (nno) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: Is this always the case, even for older versions of Matlab and Octave?
Is there any reason not to use sprintf('\n')? That would seem more readable to me.

**Comment by nno** [comment]
Very good, thanks. Please be aware that I really appreciate your efforts.

If you could make the prettify printing optional, that would be great (this reminds me that I should try to write some tests for that part as well; and testing the simple cases will be easier).

File: MOxUnit/assertEqual.m (hunk 6 above)
Comment (scottclowe) [inline MOxUnit/assertEqual.m]: Ok, but know that they are wrong about this.

Removed from PR.

File: MOxUnit/moxunit_util_elem2str.m (hunk 11 above)
Comment (scottclowe) [inline MOxUnit/moxunit_util_elem2str.m]: Removed from PR.

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 4 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: I think it needs a lookback.
The easy fix is a second regex replacement which searches for `(^|[^\\])(<>)`. This is inelegant, but very easy.

Comment (nno) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: How about adding this in a separate function `moxunit_util_strip_xml`? 
If you could add a unit test that tests a couple of inputs that would be even better!

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 3 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: Now fixed.

File: MOxUnit/@MOxUnitTestResult/report_xml.m (hunk 4 above)
Comment (scottclowe) [inline MOxUnit/@MOxUnitTestResult/report_xml.m]: I can do that.

File: MOxUnit/moxunit_runtests.m (hunk 8 above)
Comment (scottclowe) [inline MOxUnit/moxunit_runtests.m]: Fixed.

File: MOxUnit/moxunit_runtests.m (hunk 9 above)
Comment (scottclowe) [inline MOxUnit/moxunit_runtests.m]: Fixed.

**Comment by scottclowe** [comment]
I would prefer it if it were possible to output both a verbose output to screen and an XML file, or indeed to output both a log file containing verbose output and an XML file. Incidentally, I was trying to make the input format mimic [py.test](https://pytest.org/latest/usage.html#creating-junitxml-format-files).

However, I do think it would ideally be better if the code were refactored so it doesn't have to save the XML string for every testcase at the time. The XML output could get very large, so it would be best to only write it to disk and never cache it. The JUnit XML files I have seen from Python include a large chunk of test file included verbatim.

I think we should cannibalise `show_non_successes` and turn it into:
- a general purpose single-error report function which comes in quite verbose and very verbose versions.
   - quite verbose shows the full error stack, as we have at the moment
   - and very verbose would use the filename and line number to copy out a section of the problem file in context.
- a wrapper for `show_non_successes` which just loops over the errors and calls this new function
With this, we could call the inner function from `show_non_successes` for XML generation. This should be called for each of the tests in the order they occurred (since that makes more sense) regardless of status, instead of all the successes, then all the fails, etc.

**Comment by nno** [comment]
So it seems that you would like to include:

1) support for both xml and raw text output
2) present the context of a file

Re (1), in any case, I agree that it would make sense that output is not cached but written directly to a file or the terminal. One approach would be to be able to store multiple file identifiers (streams) in the TestResult object; these could be stored in a struct to mimick a dictionary. For example, s.streams.text and s.streams.xml could be stored in TestResult. 
For showing a summary of the failed tests: would that be *after* all tests have been run? Your proposal seems to propose to do it while running all the tests; I like the possibility to just show progress while running the text, and show verbose details of failed / erroring tests *after* all tests have been run. So that would be done by `show_non_successes`. Does that make sense to you?

Re (2), we could store line number in the test case. This would require adding this to moxunit_util_mfile_subfunctions (so that it returns line numbers) which is then used to make a FunctionHandleTestCase; and the possibility to either query these; or to provide the context for a test. How exactly would you like to show context? The entire (sub)function in a file?

It may be good to tackle these two things separately. How about first thinking about adding xml output support before worrying about showing context?

**Comment by scottclowe** [comment]
1a) No, I would also like to generate the XML after running all the tests.
It is currently cached just because that was easier me to write, in
particular to write without disturbing so much pre-existing code.
1b) I feel that `show_non_successes` is doing two things at the moment:
generating test for a single failure and looping through all the failures
of one type. I would like to see it changed to be two functions, one
`show_single_non_success` and another `show_all_non_success`. After it is
divided in this way, `show_single_non_success` can be used to inject the
contents of the `<failure></failure>` tag of the XML. Currently it can't be
used for this just because it spews out every non-success all at once. I
think this would be better than changing `show_non_successes` to have a
flag to tell it whether or not to inject XML. One could make
`show_single_non_success` a subfunction of `show_all_non_success` and
have `show_all_non_success`,
however in this scenario `report_xml` should also be a subfunction of
`show_all_non_success`,
since it is `report_xml` which I wish to call `show_single_non_success`.
Does that make sense? I'm not sure its easy to explain what I mean here.

2a) The line number is stored in the error stack, which we already have
stored, so there is no need to store the line number separately.
2b) It would be initially be easier to show a range of lines around the
error line. Afterwards it could be increased to the entire subfunction of
the failing testcase.

Yes, adding the context would definitely be in a different PR. This PR
indeed should just be for adding XML support. :)

On 7 December 2015 at 16:35, Nikolaas N. Oosterhof <notifications@github.com

**Comment by nno** [comment]
I'm currently working on some refactoring work by adding new TestOutcome classes; hopefully later today I will be able to send a new PR. Support for XML should be quite easy to add after the refactoring.

**Comment by scottclowe** [comment]
Okay, sounds good.

**Comment by nno** [comment]
PR #13 now aims to implement XML output support.

**Comment by scottclowe** [comment]
Is there anything left in this PR which should be used, or is it now all redundant?

**Comment by nno** [comment]
@scottclowe I think the PR has become redundant now. 
Yet, thanks a lot for your work on the PR and the useful comments on #13 - these have improved the code significantly. I've acknowledged you in some of the commits.
